### ISSUE TITLE
[Question] Possible to use MPC using external simulator's states?

### ISSUE THREAD

**Issue by hocaso** [do-mpc/do-mpc#446]
Is it possible to use your library along with my own states (obtained from my own simulator) instead of do-mpc's simulator? I successfully implemented a quadrotor stabilization model following do-mpc's examples and simulator, but I would like to use my own sim and states that differ a little bit. The differential equations are exactly the same, but the rest like the integrator or computation precision could be different.

Naively I thought on feeding my x state in the iteration loop when calculating the action like this:

```python
my_x = my_sim.init()
for i in range(30):
    u = mpc.make_step(x0=my_x)
    my_x = my_sim.step(u)
```

but it does not work and states goes to infinity (for each step the error increases). Logically I think it should work, since it's like setting the initial value of the `mpc.x0` to mine (but instead of only at the beginning, in each step). Maybe there is a hidden info in the mpc object that expect do-mpc's simulator state instead of another (like the history or dot_x)?

As an another example, I just used do-mpc simulator's state, but after obtaining it I have normalized the quaternion inside and used that state for the next action. In this case too the state goes to infinity.

If there is a way to do this, could you show me how?

**Comment by fremk** [comment]
Hello,
did you have any luck figuring this out?
I am also interested in using my own simulator that simulates the states based on the command generated by the mpc controller. Although I didn't start implementing yet but your issue seems to be an important one to address in case I wanted to use this library.
Best,
Karim

**Comment by hocaso** [comment]
Yes, I tried resetting history for every step without success. Maybe do-mpc uses the previous state for faster optimization and clashes with my own state? Don't know how to feed it correctly though.

For your knowledge, if the system is simple/stable or depending on your simulator, you can directly use do-mpc's simulator/actions and it will work. I tried with both 2 and 3 dimensional quadrotor and the 2D version works.

In the end, I ended up implementing the controller with Casadi library directly. Although I'm guessing it's much simpler than do-mpc's, I have full control over what type of optimization and integration is implemented.

**Comment by chaveza9** [comment]
I think that you can avoid using the simulator altogether. Just use mpc.make_step(x0)  and to recover the state you can use something like mpc.data.prediction(('_x','x'))

**Comment by JoshuaAda** [comment]
Hi,
first of all thank you for using Do-MPC. As pointed out in the discussion using the pure MPC controller and a different simulator should be possible as they are not interconnected. Typically, just using mpc.make_step(x0=my_x) should work fine. A few suggestions from my side:

- As already proposed mpc.reset_history() should reset the MPC controller. Note that the mpc controller is initialised in the normal case with the past my_x and depending on your optimization problem with your past input. You will lose that once you reset.
- Before using mpc.make_step(x0=my_x), you can set mpc.x0=my_x and mpc.u0=u, this should initialise the MPC with your current my_x state
- Finally, please check carefully whether your simulation model is really the same as the model that is used for the optimization. Even a slight different time discretization can lead to the problems you have described above.
If the problem still persist you can contact us again maybe with some code example where you belive the mpc controller is not working satisfactory with an external simulator.
Hope this helps,
Your Do-MPC Team

**Comment by hocaso** [comment]
Hello,
Thank you for your response, and I apologize for my delayed reply, I could not find time to analyze it properly until now.
Based on your suggestions, I think this general workflow is better than the previous one for my case?

```
def step(self, x, u_prev):
  self.mpc.reset_history()
  self.mpc.u0 = u_prev
  self.mpc.x0 = x
  u = self.mpc.make_step(x0=x)
  return u[:,0]
```

However, the issue persists as it still diverges. I have confirmed that all differential equations and parameters, such as the simulation's dt, remain the same. To ensure the same integrator is being used, could you clarify if do-mpc employs a simple RK4 method or a more complex approach? I don't recall seeing this information in the documentation.

To be honest, I'm somewhat unsure about what else to investigate, so I'm going to try to implement all my program in a single file to facilitate sharing and testing.

What I find strange is that in the CasADi version I've developed, changing the integration method does not hinder convergence to the optimal state. Although I have used multiple shooting method instead of collocation, so I guess it is not comparable.

**Comment by hocaso** [comment]
I'm going to close the issue, as I'm not completely sure that all my formulas were correct. If I encounter the error again, I will make sure that I verify the code and reopen the issue.
